UIA handler/notification event: handle elements without window handle property set such as window visual state announcements from File Explorer/Windows shell

josephsl opened the pull request:
### Link to issue number:
Closes #18220 
Closes #17841 

### Summary of the issue:
NVDA does not announce window visual states such as restore and snap when Windows+arrow keys are pressed.

### Description of user facing changes:
NVDA will announce window visual states (for the foreground window) when Windows+arrow keys are pressed.

### Description of developer facing changes:
Window state notifications are raised by a Windows shell (File Explorer) element without a native window handle. Therefore, allow notifications to be processed if the app module says yes, and if so, use foreground window handle as "window handle" for the element (because UIA NVDA objects require a window handle during construction even though all that will be done is process events).

### Description of development approach:
In UIA handler/notification event handler:

* If native window handle is not found (including traversing the UIA tree back to the root element), ask the app module for the process (where the element lives) if notification event should be processed.
* If the app module says yes, set foreground window handle as "window handle" for the UIA element.

Base app module:

* Added a new method, "shouldProcessUIANotificationEventNoWindowHandle" to inform UIA handler if UIA notifications for elements without native window handle should be processed.
* The new method will return False by default in line with previous behavior of dropping notification events for elements without window handle property set.

File Explorer app module:

* Say "yes" to "shouldProcessUIANotificationEventNoWindowHandle" call if window snap activity Id is being handled.
* Added app module version of UIA notification event handler to announce window visual state/snap notification from everywhere.

### Testing strategy:
Manual testing: from any app window (such as a web browser), press Windows+arrw keys to snap windows to part of the screen or press Windows+up arrow to maximize the foreground window.

Prior to this PR: no window state announcement.
After this PR: NVDA will announce window snap notification.

### Known issues with pull request:
While the foreground window handle is an obvious choice (as the nanouncemenet pertians to window visual state changes), there might be a better way such as a custom UIA property. Ideally, window handle should be exposed by this shell element and hopefully Microsoft can resolve this. Note that File Explorer is not the only app with this issue - Windows 11 Voice Access exhibits the same problem (#16862).

### Code Review Checklist:

<!--
This checklist is a reminder of things commonly forgotten in a new PR.
Authors, please do a self-review of this pull-request.
Check items to confirm you have thought about the relevance of the item.
Where items are missing (eg unit / system tests), please explain in the PR.
To check an item `- [ ]` becomes `- [x]`, note spacing.
You can also check the checkboxes after the PR is created.
A detailed explanation of this checklist is available here:
https://github.com/nvaccess/nvda/blob/master/projectDocs/dev/githubPullRequestTemplateExplanationAndExamples.md#code-review-checklist
-->

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [x] Security precautions taken.

<!-- Please keep the following -->
@coderabbitai summary

josephsl committed:
UIA handler: ask app modules if they should accept UIA notifications from elements without window handles. Re #18220.

UIA notification event fails if UIA element has no native window handle. However, some of these elements report system information such as window visual state (restore/maximize/snap) yet do not set native window handle value. Therefore, ask app modules if notifications from these elements should be accepted (notable case: File Explorer/window visual states).

josephsl committed:
App module handler/AppModule class: define shouldProcessUIANotificationEventNoWindowHandle method to determine if notifications should be received from elements without hwnd set. Re #18220.

Similar to UIA propewrty change event handler: add a method in the base app module to say yes or no if asked to process UIA notification events for elements (senders) without window handle defined. Method signature is same as UIA notification event except arguments other than sender (element) are keyword arguments - not all app modules may wish to work with all UIA notification event arguments. By default, this method will return False for compatibility with previous NVDA behavior of dropping notification events without window handle set.

josephsl committed:
UIA handler/notification event handler: set foreground window handle as 'widow handle' if native window handle is not set for an element. Re #18220.

UIA NVDA objects need window handle as part of its construction, including while processing events. However, some elements with important notifications (notably window visual state notifications) do not set native window handle property. Therefore, set foreground window handle as the element's 'window handle' to allow notification event to propagate across handlers (global plugins, app modules, NVDA object).

josephsl committed:
appModules/explorer: update copyright header

josephsl committed:
File Explorer app module: announce window visual state notifications from everywhere. Re #17841, #18175.

The element that reports window visual state changes (restore/masimize/minimize/snap) does not define a native window handle property. Therefore, let File Explorer app module say 'yes' when asked to process UIA notification event from this element. Note that this is not the only time a 'no window handle notification' is used - the other case is input layout switch notifications (say, changing between input languages), thus limit processing to snap anouncements.

josephsl committed:
Changelog: add 2025.2/new features entry on window visual state announcement (Windows 11 24H2 and Server 2025 and later). Re #17841, #18175, #18220

josephsl committed:
UIA handler: correct the NVDA issue reference

pre-commit-ci[bot] committed:
Pre-commit auto-fix

josephsl committed:
UIA handler/notification event handler debug lo: spacing

josephsl commented on source/appModules/explorer.py:
```diff
@@ -560,3 +561,28 @@ def event_UIA_elementSelected(self, obj: NVDAObject, nextHandler: Callable[[], N
 			)
 		nextHandler()
+
+	def shouldProcessUIANotificationEventNoWindowHandle(
```
Ah, that's taken care of at the UIA handler event handler stage when the event handler fails to locate native window handle. You're right about making this a more generic method (I think we should also document "shouldProcessUIAPropertyChangeEvent" (done by Jamie) somewhere.

josephsl committed:
UIA handler/notification event handler and app modules: shouldProcessNOtificationEventNoWindowHandle -> shouldProcessUIANotificationEvent (generalizability). Re #18220, #17841.

Rename the app module method to make it more generalizable (including handling events from elements with no window handle set).

josephsl committed:
App modules: say 'no' if nearest window handle returns None. Re #18220.

Nearest window handle getter from UIA handler can say 'None' if window is not found, and calling bool(None) gives False. Therefore, use the boolean transformation of the nearest window handle call as base implementation for 'shouldProcessUIANotificationEvent' method.

josephsl committed:
File Explorer app module: call the parent method after handling specific cases when asked to process UIA notification event. Re #17841, #18220

josephsl committed:
UIA handler: update source code comments (elements without native window handle can report notifications only if app modules say so). Re #17841, #18220

josephsl commented on source/appModules/explorer.py:
```diff
@@ -560,3 +561,28 @@ def event_UIA_elementSelected(self, obj: NVDAObject, nextHandler: Callable[[], N
+		self,
+		sender,
+		activityId: str = "",
+		**kwargs,
```
I see. We have precedence for that: modern keyboard and Windows 11 Calculator app modules handle notification events with keyword arguments. Looks like I need to look for an alternative then (any suggestions?).

josephsl commented on source/appModules/explorer.py:
```diff
@@ -560,3 +561,28 @@ def event_UIA_elementSelected(self, obj: NVDAObject, nextHandler: Callable[[], N
+	) -> bool:
+		if activityId == "Windows.Shell.SnapComponent.SnapHotKeyResults":
+			return True
+		return super().shouldProcessUIANotificationEvent(sender, **kwargs)
```
Good catch.

josephsl committed:
appModules/explorer: spell out argument names for shouldProcessUIANotificationEvent/event_UIA_notification instead of using kwargs dictionary. Re #18220.

Reviewed by Leonard de Ruijter: spell out argument names for shouldProcessUIANotificationEvent and event_UIA_notification to make the code more readable.

josephsl committed:
Changelog: formatting. Re #17841, #18175.

NOted by Cyrille Bougot: add grave accent (1) to emphasize hotkeys.

Co-authored-by: Cyrille Bougot <cyrille.bougot2@laposte.net>

josephsl committed:
Merge branch 'master' into i18220UIANotificationNoWindowHandle

SaschaCowley commented on source/UIAHandler/__init__.py:
```diff
@@ -1041,10 +1041,42 @@ def IUIAutomationNotificationEventHandler_HandleNotificationEvent(
+						f"at request of appModule {appMod.appName}",
+					)
+				return
+		# Take foreground window handle as a substitute if window handle is not set.
```
Why foreground window? Wouldn't it make more sense to use the desktop window, which all windows are children of?

LeonarddeR replied:
Shouldn't the used window handle at least belong to the same app/process?

josephsl replied:
At least for window visual state reporting, no - that notification comes from File Explorer (Windows shell), not from the foreground app.

SaschaCowley reviewed (changes requested):
Note that @michaelDCurran has been requested to review this as well due to my inexperience with UIA

josephsl commented on source/appModuleHandler.py:
```diff
@@ -765,6 +765,28 @@ def shouldProcessUIAPropertyChangedEvent(self, sender, propertyId):
 
+	def shouldProcessUIANotificationEvent(
+		self,
+		sender,
```
This will involve importing UIAHandler.UIA or equivalent, meaning the app module handler module will rely on UIA support which I don't think should be done at this time (the app module handler uses window handles and other base elements for various operations, including UIA good/bad window responses from app modules). Besides, UIA handler imports app module handler, and if importing UIA handler from app module handler, this may introduce circular dependency (the solution would be to import the just needed attributes). Thanks.

SaschaCowley replied:
For future reference, you can use [`typing.TYPE_CHECKING`](https://docs.python.org/3/library/typing.html#typing.TYPE_CHECKING) to conditionally import things only for the purpose of static type analysis. For instance:

```py
from typing import TYPE_CHECKING

if TYPE_CHECKING:
	from someModule import SomeClass

def function(
	argument: "SomeClass",
) -> None:
	...
```

josephsl committed:
Merge branch 'master' into i18220UIANotificationNoWindowHandle

josephsl committed:
appModules/explorer: add missing comma to notification event handler call. Re #18220

josephsl committed:
App modules (base and File Explorer): add type hints to 'sender' argument in notification event handler. Re #18220.

Note from Sascha Cowley (NV ACcess): add type hints please (do this in app modules by importing UIA client from COM interfaces).

josephsl committed:
Base app module: add parameter comments to 'shouldProcessUIANotificationEvent' method. Re #18220.

Note from Sascha Cowley (NV Access): add parameter hints to notification processing event to improve docstring clarity.

josephsl committed:
UIA handler: use desktop window handle when assigning 'window handle' to UIA elements without window handles. Re #18220.

NOte from reviewers: would it be better to use desktop window handle rather than foreground handle? The issue: some UIA elements do not connect back to the desktop through UIA tree traversal. The foreground window handle makes sense when handling the primary case of window visual state announcement, but there are other places where UIA notification events without window handles (at least from NVDA's perspective) will be seen such as Voice Access (Windows 11). Therefore, switch to using desktop window handle to communicate the need to handle notification events more generically.

michaelDCurran reviewed (approved):
Approving, at least on the approach.

josephsl commented:
Hi, thanks. I think we still need to persuade Microsoft to take UIA tree management more seriously to avoid a situation like this. I’ll introduce Windows 11 Voice Access support within 48 hours of PR merge.

SaschaCowley committed:
Fix up docstring

SaschaCowley committed:
Merge branch 'master' into i18220UIANotificationNoWindowHandle

# Conflicts:
#	user_docs/en/changes.md

pre-commit-ci[bot] committed:
Pre-commit auto-fix
